ext/multicore/rpmsg_lite/lib/rpmsg_lite/porting/platform/lpc5411x/rpmsg_platform_zephyr_ipm.c

```diff
@@ -0,0 +1,265 @@
+ * All rights reserved.
+ *
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted (subject to the limitations in the disclaimer below) provided
+ * that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY THIS LICENSE.
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <assert.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "rpmsg_platform.h"
+#include "rpmsg_env.h"
+#include <ipm.h>
+
+static int isr_counter = 0;
+static int disable_counter = 0;
+static void *lock;
+static struct device *ipm_handle = NULL;
+
+void platform_ipm_callback(void *context, u32_t id, volatile void *data)
+{
+    if ((*(uint32_t*)data) & 0x01)
+    {
+        env_isr(0);
+    }
+    if ((*(uint32_t*)data) & 0x02)
+    {
+        env_isr(1);
+    }
+}
+
+int platform_init_interrupt(int vq_id, void *isr_data)
+{
+    /* Register ISR to environment layer */
+    env_register_isr(vq_id, isr_data);
+
+    env_lock_mutex(lock);
+
+    assert(0 <= isr_counter);
+
+    isr_counter++;
+
+    env_unlock_mutex(lock);
+
+    return 0;
+}
+
+int platform_deinit_interrupt(int vq_id)
+{
+    /* Prepare the MU Hardware */
+    env_lock_mutex(lock);
+
+    assert(0 < isr_counter);
+    isr_counter--;
+    if ((!isr_counter) && (ipm_handle != NULL))
+    {
+        ipm_set_enabled(ipm_handle, 0);
+    }
+
+    /* Unregister ISR from environment layer */
+    env_unregister_isr(vq_id);
+
+    env_unlock_mutex(lock);
+
+    return 0;
+}
+
+void platform_notify(int vq_id)
+{
+    switch (RL_GET_LINK_ID(vq_id))
+    {
+        case 0:
```

Comment:
Why is only link ID 0 supported here?

Reply:
LPC54114 has two cores so there is an expectation that most of the applications would use a single link there. I am going to replace 0 with RL_PLATFORM_LPC5411x_M4_M0_LINK_ID. If somebody wants to use multiple instances of RPMsg-Lite, he or she has to define more link IDs and implement it in the switch statement.